pkg/arvo/app/chat-hook.hoon

```diff
@@ -157,14 +181,57 @@
 ++  peer-mailbox
   |=  pax=path
   ^-  (quip move _this)
+  ?>  ?=(^ pax)
+  =/  last  (dec (lent pax))
+  =/  backlog-start=(unit @ud)
+    %+  rush
+      (snag last `(list @ta)`pax)
+    dem:ag
+  =>  .(pax `path`(oust [last last] `(list @ta)`pax))
   ?>  ?=([* ^] pax)
   ?>  (~(has by synced) pax)
   ::  scry permissions to check if read is permitted
   ?>  (permitted-scry [(scot %p src.bol) %chat (weld pax /read)])
   =/  box  (chat-scry pax)
   ?~  box  !!
   :_  this
-  [ost.bol %diff %chat-update [%create (slav %p i.pax) pax]]~
+  :-  [ost.bol %diff %chat-update [%create (slav %p i.pax) pax]]
+  ?:  ?&(?=(^ backlog-start) (~(got by allow-history) pax))
+    (paginate-messages pax u.box u.backlog-start)
+  ~
+::
+++  paginate-messages
+  |=  [=path =mailbox start=@ud]
+  ^-  (list move)
+  =/  moves=(list move)  ~
+  =/  end  (lent envelopes.mailbox)
+  ?:  |((gte start end) =(end 0))
+    moves
+  =.  envelopes.mailbox  (slag start `(list envelope)`envelopes.mailbox)
+  |-  ^-  (list move)
+  ?~  envelopes.mailbox
+    moves
+  ?:  (lte end 5.000)
```

Comment:
5000 is a lot, have you tested that many to make sure it doesn't bail:meme?

Reply:
Works just fine, just tested it.